packages/mdc-auto-init/index.ts

```diff
@@ -91,6 +91,7 @@ export function mdcAutoInit(root = document, warn = CONSOLE_WARN) {
       writable: false,
     });
     components.push(component);
+    node.setAttribute(AUTO_INIT_STATE_ATTR, INTIALIZED_STATE);
```

Comment:
I feel, we can take this further by moving it to MDC base component (`MDCComponent`). Auto init function then look for elements that do not have this attribute.

So this'll also avoid problem of initialing a component twice.

WDYT?

Reply:
The only component that will be using this is the auto-init package, correct? What will other components do with the `AUTO_INIT_STATE_ATTR`?

Reply:
Right, we need to name this attribute more generic.

`data-mdc-init-state="initialized"`?

Reply:
MDCComponent can throw an error when component is initialized twice on same DOM element. Where as, autoInit picks up the elements that do not have this initialized data attribute and initializes them.

Reply:
I see - that's a good idea

Reply:
Well actually - MDC Component should early exit if its already been initialized. Right? I don't think we need to throw an error.

Reply:
IMO, the component should not be forgiving when it is initialized more than once. Throwing error explicitly makes developer discover the bug easier.

Reply:
Also, currently there is no way to get component instance from DOM element reference. I think It would be great addition if we can also access it via MDCComponent API (example, `MDCComponent.getInstance(menuEl)`).

Suggestion: MDCComponent can store component instance globally (example, `window.__mdcComponentInstanceXXXX` where `XXXX` is uniq id), and store this string to data attribute of DOM element where the component is initialized.

```html
data-mdc-component-id="__mdcComponentInstanceXXXX"
```

This can potentially replace `init-status` data attribute.

WDYT?

Reply:
What about a warning instead of an error? Throwing an error might cause the user's app to crash or become in a bad state. 

RE storing DOM Node on component: I think is a feature separate from this. I think we could add whatever you want on top of what we're currently building in this PR.

Reply:
Idea is to not allow multiple instance of same component on same DOM element.

```ts
new MDCFoo(fooEl); // Always expected to return an instance of `MDCFoo`
```

Silently ignoring it or throwing a warning does not look semantic to me.

Reply:
Gotcha - ok that makes sense. I'll throw a warning. 

So you're ok with storing DOM Node in separate PR?

Reply:
I just tried setting the attribute in the constructor of mdc-base/component like so:

```
  constructor(
      root: Element,
      foundation?: FoundationType,
      ...args: Array<unknown>
  ) {
    this.root_ = root;
    debugger
    if (this.root_.getAttribute(INIT_STATE_ATTR) === INTIALIZED_STATE) {
      throw new Error('MDCComponent: component is already initialized.');
    }
    this.initialize(...args);
    this.root_.setAttribute(INIT_STATE_ATTR, INTIALIZED_STATE);
```

This however causes an issue. In checkbox (and others), we instantiate both the ripple and checkbox on the same .mdc-checkbox element. So when the checkbox initializes the ripple, it throws the error. I think we could get around this using a boolean when instantiating the ripple within the checkbox (and others). I wanted to run this by you before moving forward.

Reply:
created https://github.com/material-components/material-components-web/issues/4715